google-cloud-firestore/lib/google/cloud/firestore/convert.rb

```diff
@@ -0,0 +1,546 @@
+                path.map { |p| escape_field_path p }.join(".")
+              else
+                path
+              end
+            end
+
+            # return a new hash and paths
+            [Hash[new_pairs.compact], paths]
+          end
+
+          def identify_leaf_nodes hash
+            paths = []
+
+            hash.map do |key, value|
+              if value.is_a? Hash
+                nested_paths = identify_leaf_nodes value
+                nested_paths.each do |nested_path|
+                  paths << "#{escape_field_path(key)}.#{nested_path}"
+                end
+              else
+                paths << escape_field_path(key)
+              end
+            end
+
+            paths
+          end
+
+          def select_by_field_paths hash, field_paths
+            new_hash = {}
+            field_paths.map do |field_path|
+              selected_hash = select_field_path hash, field_path
+              # new_hash = deep_merge_hashes new_hash, selected_hash
+              deep_merge_hashes new_hash, selected_hash
+            end
+            new_hash
+          end
+
+          def select_field_path hash, field_path
+            ret_hash = {}
+            tmp_hash = ret_hash
+            prev_hash = ret_hash
+            dup_hash = hash.dup
+            nodes = unescape_field_path field_path
+            last_node = nil
+
+            # squash nodes until the key exists?
+            until dup_hash.key? nodes.first
+              nodes.unshift "#{nodes.shift}.#{nodes.shift}"
+              break if nodes.count <= 1
+            end
+
+            nodes.each do |node|
+              prev_hash[last_node] = tmp_hash unless last_node.nil?
+              last_node = node
+              tmp_hash[node] = {}
+              prev_hash = tmp_hash
+              tmp_hash = tmp_hash[node]
+              dup_hash = dup_hash[node]
+            end
+            prev_hash[last_node] = dup_hash
+            ret_hash
+          end
+
+          def deep_merge_hashes left_hash, right_hash
+            right_hash.each_pair do |key, right_value|
+              left_value = left_hash[key]
+
+              if left_value.is_a?(Hash) && right_value.is_a?(Hash)
+                left_hash[key] = deep_merge_hashes left_value, right_value
+              else
+                left_hash[key] = right_value
+              end
+            end
+
+            left_hash
+          end
+
+          START_FIELD_PATH_CHARS = /\A[a-zA-Z_]/
+          INVALID_FIELD_PATH_CHARS = /[\~\*\/\[\]]/
+          ESCAPED_FIELD_PATH = /\A\`(.*)\`\z/
+
+          def extract_field_paths hash
+            field_paths = hash.keys.map do |path|
+              unless path.is_a? Array
+                path = path.to_s.split(".")
+                path.each do |p|
+                  # Only check for invalid characters when a string
+                  if INVALID_FIELD_PATH_CHARS.match(p) || p["."]
+                    fail ArgumentError, "invalid character"
```

Comment:
Can you tell the users here to use the FieldPath class?

Reply:
Updated.